Add template as_timedelta

eifinger opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

As discussed in https://github.com/home-assistant/architecture/discussions/753 this adds the template `as_timedelta` which parses durations in the form of `PT10M` to a `timedelta` object.
Among other usecases Alexa provides durations in this ISO8601 format.

I verified it works with the following automation:

```yaml
alias: New Automation
description: ''
trigger:
  - platform: event
    event_type: parse_iso8601_duration
condition: []
action:
  - delay: '{{ as_timedelta("PT10S") }}'
  - service: persistent_notification.create
    data:
      message: iso8601
mode: single
```

I added isodate as a dependency rather than extracting the functionality as this would mean ~400 lines of code taken from [here](https://github.com/gweis/isodate/blob/8856fdf0e46c7bca00229faa1aae6b7e8ad6e76c/src/isodate/isoduration.py#L60-L73), [here](https://github.com/gweis/isodate/blob/8856fdf0e46c7bca00229faa1aae6b7e8ad6e76c/src/isodate/isoduration.py#L88-L109), [here](https://github.com/gweis/isodate/blob/8856fdf0e46c7bca00229faa1aae6b7e8ad6e76c/src/isodate/isodatetime.py#L24-L33), [here](https://github.com/gweis/isodate/blob/8856fdf0e46c7bca00229faa1aae6b7e8ad6e76c/src/isodate/isotime.py#L16-L144), [here](https://github.com/gweis/isodate/blob/8856fdf0e46c7bca00229faa1aae6b7e8ad6e76c/src/isodate/isodates.py#L15-L192), [here](https://github.com/gweis/isodate/blob/8856fdf0e46c7bca00229faa1aae6b7e8ad6e76c/src/isodate/isotzinfo.py#L18-L32) and [here](https://github.com/gweis/isodate/blob/8856fdf0e46c7bca00229faa1aae6b7e8ad6e76c/src/isodate/tzinfo.py#L57-L98)

If we want to avoid the dependency I can extract and add the ~400 lines to e.g. `helpers/iso8601.py`

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/22740

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

eifinger committed:
Add template parse_iso8601_duration

balloob commented on homeassistant/package_constraints.txt:
```diff
@@ -18,6 +18,7 @@ hass-nabucasa==0.54.0
 home-assistant-frontend==20220504.1
 httpx==0.22.0
 ifaddr==0.1.7
+isodate==0.6.1
```
Meh, I don't like adding a new dependency. It's pulling in more code and only for a template function. The code you're linking is showing timezone parsing, but that's not part of a duration at all ?

eifinger replied:
Thanks for the feedback. I will dive in and pull out the good stuff.

eifinger committed:
Implement simple iso8601 duration

eifinger committed:
Correct iso8601 duration import

eifinger commented:
I extracted the regex and implemented an simple version for ISO8601 Duration parsing which does not support years months and negative durations as well as alternative representations.

If we need a fullblown implementation [pendulum](https://github.com/sdispater/pendulum/blob/f1df7dc3f838bd4ab1075ba25c8b6ce5d8141995/pendulum/parsing/iso8601.py#L260) seems a more clean implementation than isodate.

eifinger commented:
@balloob codecov is failing for function comments. Otherwise this is ready for review =)

frenck commented:
Should we just has an `as_duration` or `as_timedelta` supporting multiple options (including iso8601) instead? Similar to how we have `as_datetime` right now.

It feel like `parse_iso8601_duration` is accurate, but feels out of bands with the existing  function names/capabilities.

eifinger committed:
Rename to as_timedelta

eifinger commented:
I renamed the template to `as_timedelta` and found an implementation [in the django codebase](https://github.com/django/django/blob/main/django/utils/dateparse.py). As we are already using their `parse_datetime` function it felt more natural to use it.

eifinger committed:
fix mypy

eifinger commented:
Is there a way to teach codecov about the copyright comments?

frenck reviewed (approved):
> Is there a way to teach codecov about the copyright comments?

Yup, ignore it completely :)

---

This is **awesome**! 🎉 

Thanks, @eifinger ❤️
